torchvision/models/regnet.py

```diff
@@ -0,0 +1,785 @@
+        )
+
+        return zip(
+            stage_widths, strides, stage_depths, group_widths, bottleneck_multipliers
+        )
+
+    @staticmethod
+    def _adjust_widths_groups_compatibilty(
+            stage_widths: List[int], bottleneck_ratios: List[float],
+            group_widths: List[int]) -> Tuple[List[int], List[int]]:
+        """
+        Adjusts the compatibility of widths and groups,
+        depending on the bottleneck ratio.
+        """
+        # Compute all widths for the current settings
+        widths = [int(w * b) for w, b in zip(stage_widths, bottleneck_ratios)]
+        group_widths_min = [min(g, w_bot) for g, w_bot in zip(group_widths, widths)]
+
+        # Compute the adjusted widths so that stage and group widths fit
+        ws_bot = [_make_divisible(w_bot, g) for w_bot, g in zip(widths, group_widths_min)]
+        stage_widths = [int(w_bot / b) for w_bot, b in zip(ws_bot, bottleneck_ratios)]
+        return stage_widths, group_widths_min
+
+
+class RegNet(nn.Module):
+    def __init__(self, params: RegNetParams) -> None:
+        super().__init__()
+
+        activation = params.activation(inplace=True)
+
+        # Ad hoc stem
+        self.stem = params.stem_type(
+            3,  # width_in
+            params.stem_width,
+            params.bn_epsilon,
+            params.bn_momentum,
+            activation,
+        )
+
+        current_width = params.stem_width
+
+        self.trunk_depth = 0
+
+        blocks = []
+        for i, (
+            width_out,
+            stride,
+            depth,
+            group_width,
+            bottleneck_multiplier,
+        ) in enumerate(params.get_expanded_params()):
+            blocks.append(
+                (
+                    f"block{i+1}",
+                    AnyStage(
+                        current_width,
+                        width_out,
+                        stride,
+                        depth,
+                        params.block_type,
+                        params.bn_epsilon,
+                        params.bn_momentum,
+                        activation,
+                        group_width,
+                        bottleneck_multiplier,
+                        params.se_ratio,
+                        stage_index=i + 1,
+                    ),
+                )
+            )
+
+            self.trunk_depth += blocks[-1][1].stage_depth
+
+            current_width = width_out
+
+        self.trunk_output = nn.Sequential(OrderedDict(blocks))
+
+        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
+        self.fc = nn.Linear(in_features=current_width, out_features=params.num_classes)
+
+        # Init weights and good to go
+        self._init_weights()
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = self.stem(x)
+        x = self.trunk_output(x)
+
+        x = self.avgpool(x)
+        x = x.flatten(start_dim=1)
+        x = self.fc(x)
+
+        return x
+
+    def _init_weights(self) -> None:
```

Comment:
I propose a different semi-standard name for this function: `reset_parameters()` + making it public for resetting the parameters if needed

Reply:
You are right to say right now we don't offer a good way to achieve this and that we probably should create a mechanism. I believe you have create a feature request for this at #3410.

This is worth discussing separately from this PR because the specific idiom proposed is not currently used in vision. In practice in all cases, the models initialize their weights inside the constructor.

Reply:
A related discussion is at https://github.com/pytorch/pytorch/issues/29523

Reply:
Guidance here also proposes to always using nn.init methods: https://pytorch.org/tutorials/prototype/skip_param_init.html#updating-modules-to-support-skipping-initialization

Reply:
(also I found that VGGNet uses `_initialize_weights`)